astropy/coordinates/baseframe.py

```diff
@@ -709,8 +856,17 @@ def transform_to(self, new_frame):
         if self._data is None:
             raise ValueError('Cannot transform a frame with no data')
 
+        if (getattr(self.data, 'differentials', None) and
+           hasattr(self, 'obstime') and hasattr(new_frame, 'obstime') and
+           np.all(self.obstime == new_frame.obstime)):
```

Comment:
> Hmm, this surely has to be np.any(self.obstime != new_frame.obstime). And that this should be tested!

Good point! I recall writing a test for this, but I don't see it now.  It's possible I did not commit it ... (it was ~1am local time, so I will blame that...)

> Also, just to be sure, since ICRS does not have an obstime, does this mean transforming to it will always work?

No, because in most (all?) cases, to get to ICRS from something that has an `obstime` you first have to do a transform into J2000 obstime.  Certainly the `AltAz` (the most important case) works that way.